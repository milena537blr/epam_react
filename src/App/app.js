import React from "react";
// import PropTypes from 'prop-types';
import Article from "../Article/Article";
import Search from "../Search/Search";
import data from "../data/data.json";
import ErrorBoundary from "../ErrorBoundary/ErrorBoundary";
import s from "./app.module.scss";
import Logo from "../Logo/Logo";
import Box from "../Box/Box";
import Button from "../Button/Button";
import Card from "../Card/Card";

class App extends React.Component {
  constructor(props) {
    super(props);

    let cards = data.cards.map(item => {
      return {
        name: item.name,
        date: item.date,
        genre: item.genre,
        id: item.id
      };
    });

    this.state = {
      data,
      listCards: [],
      searchText: "",
      cards,
      isSearchActive: true,
      currentCard: {},
      filterText: ""
    };
  }

  handleSearchTextChange = searchText => {
    this.setState({ searchText });
  };

  handleCardClick = currentCardId => {
    this.setState({
      isSearchActive: false
    });

    this.state.data.cards.forEach(currentCard => {
      if (currentCard.id === currentCardId) {
        this.setState({ currentCard });
        return;
      }
    });
  };

  handleSearchClick = () => {
    this.setState({
      isSearchActive: true
    });
  };

  render() {
    const isSearchActive = this.state.isSearchActive;

    const cards = this.state.cards.map(card => {
      return (
        <Card
          cardClick={this.handleCardClick}
          card={card}
          key={card.id}
        />
      );
    });

    return (
      <ErrorBoundary>
        <header className={s.header}>
          <div className={s.overlay} />
          <section className={s.container}>
            <Box align="space-between" verticalAlign="middle" marginBottom={8}>
              <Logo />
              <Button
                onHandleSearchClick={this.handleSearchClick}
                text="SEARCH"
                size="large"
                color="white"
              />
            </Box>
            {isSearchActive ? (
              <Search
                searchText={this.state.searchText}
                onSearchTextChange={this.handleSearchTextChange}
              />
            ) : (
              <Article card={this.state.currentCard} />
            )}
          </section>
        </header>
        <section className={s.panel}>
          <Box
            align="space-between"
            verticalAlign="middle"
            className={s.container}
          >
            <div>7 movies found</div>
            <div>
              <span>Sort by </span>
              <span>release date </span>
              <span>rating</span>
            </div>
          </Box>
        </section>
        <main className={s.main}>
          <div className={s.container}>
            <Box className={s.wrapper} align="center">
              {cards}
            </Box>
          </div>
        </main>
        <footer className={s.footer}>
          <div className={s.container}>
            <Logo />
          </div>
        </footer>
      </ErrorBoundary>
    );
  }
}

/* App.propTypes = {
  name: PropTypes.string,
}; */

export default App;
