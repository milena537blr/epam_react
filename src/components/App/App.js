import React from "react";
import Article from "../Article/Article";
import Search from "../Search/Search";
import ErrorBoundary from "../ErrorBoundary/ErrorBoundary";
import s from "./app.module.scss";
import Logo from "../Logo/Logo";
import Box from "../Box/Box";
import Button from "../Button/Button";
import Card from "../Card/Card";
import PropTypes from "prop-types";

class App extends React.Component {
  constructor(props) {
    super(props);

    let cards = this.props.data.cards.map(item => {
      return {
        name: item.name,
        date: item.date,
        genre: item.genre,
        id: item.id
      };
    });

    this.state = {
      data: this.props.data,
      listCards: [],
      searchText: "",
      cards,
      isSearchActive: true,
      currentCard: {},
      filterText: ""
    };
  }

  handleSearchTextChange = searchText => {
    this.setState({ searchText });
  };

  handleCardClick = currentCardId => {
    this.setState({
      isSearchActive: false
    });

    this.state.data.cards.forEach(currentCard => {
      if (currentCard.id === currentCardId) {
        this.setState({ currentCard });
        return;
      }
    });
  };

  handleSearchClick = () => {
    this.setState({
      isSearchActive: true
    });
  };

  render() {
    const isSearchActive = this.state.isSearchActive;

    return (
      <ErrorBoundary>
        <header className={s.header}>
          <div className={s.overlay} />
          <section className={s.container}>
            <Box align="space-between" verticalAlign="middle" marginBottom={8}>
              <Logo />
              <Button
                onHandleSearchClick={this.handleSearchClick}
                text="SEARCH"
                area-label="Display search area"
                size="large"
                color="white"
                data-testid="search-switcher"
              />
            </Box>
            {isSearchActive ? (
              <Search
                searchText={this.state.searchText}
                onSearchTextChange={this.handleSearchTextChange}
              />
            ) : (
              <Article card={this.state.currentCard} />
            )}
          </section>
        </header>
        <section className={s.panel}>
          <Box
            align="space-between"
            verticalAlign="middle"
            className={s.container}
          >
            <div>7 movies found</div>
            <div>
              <span>Sort by </span>
              <span>release date </span>
              <span>rating</span>
            </div>
          </Box>
        </section>
        <main className={s.main}>
          <div className={s.container}>
            <Box className={s.wrapper} align="center">
              {this.state.cards.map(card => {
                return (
                  <Card
                    cardClick={() => this.handleCardClick(card.id)}
                    card={card}
                    key={card.id}
                  />
                );
              })}
            </Box>
          </div>
        </main>
        <footer className={s.footer}>
          <div className={s.container}>
            <Logo />
          </div>
        </footer>
      </ErrorBoundary>
    );
  }
}

App.propTypes = {
  data: PropTypes.object
};

export default App;
