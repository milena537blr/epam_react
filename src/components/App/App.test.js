import React from "react";
import { mount } from "enzyme";
import App from "./App.js";
import { render, fireEvent } from "@testing-library/react";

const card1 = {
  id: 1,
  date: 2007,
  description: "American drama",
  genre: "Dramas",
  name: "Vampire diaries",
  text:
    "The Vampire Diaries is an American supernatural teen drama television series developed by Kevin Williamson and Julie Plec, based on the popular book series of the same name written by L. J. Smith.",
  time: 145
};

const card2 = {
  id: 2,
  date: 2014,
  description: "Drama/Comedy-drama",
  genre: "Drama",
  name: "The Fault in Our Stars",
  text:
    "Hazel Grace Lancaster (Shailene Woodley), a 16-year-old cancer patient, meets and falls in love with Gus Waters (Ansel Elgort), a similarly afflicted teen from her cancer support group. Hazel feels that Gus really understands her. ",
  time: 133
};

describe("<App />", () => {
  const data = { cards: [card1, card2] };
  const component = mount(<App data={data} />);

  const { getByText, getByAltText, getByTestId, queryByTestId } = render(<App data={data} />);

  test("snapshot test", () => {
    expect(component).toMatchSnapshot();
  });

  test("renders all Cards", () => {
    expect(component.find("Card").length).toBe(2);
  });

  test("generates name, date, image and genre of Card", () => {
    expect(getByText(data.cards[0].name)).toBeInTheDocument();
    expect(getByText(data.cards[0].date.toString())).toBeInTheDocument();
    expect(getByText(data.cards[0].genre)).toBeInTheDocument();
    expect(getByAltText(data.cards[0].name)).toBeInTheDocument();
  });

  test("generates images for all Cards", () => {
    expect(component.find("img").length).toBe(2);
  });

  test("generates Article instead Search after Card had been clicked", () => {
    expect(queryByTestId("article")).toBeNull();
    expect(getByTestId("search")).toBeInTheDocument();
    fireEvent.click(getByAltText(data.cards[1].name));
    expect(queryByTestId("search")).toBeNull();
    expect(getByTestId("article")).toBeInTheDocument();
  });

  test("generates Search instead Article after search button had been clicked", () => {
    fireEvent.click(getByAltText(data.cards[1].name));
    fireEvent.click(getByTestId("search-switcher"));
    expect(queryByTestId("article")).toBeNull();
    expect(getByTestId("search")).toBeInTheDocument();
  });

  test("generates name, description, text, time and date in Article", () => {
    fireEvent.click(getByAltText(data.cards[0].name));
    expect(getByTestId("article")).toHaveTextContent(data.cards[0].name);
    expect(getByTestId("article")).toHaveTextContent(data.cards[0].description);
    expect(getByTestId("article")).toHaveTextContent(data.cards[0].text);
    expect(getByTestId("article")).toHaveTextContent(data.cards[0].time.toString());
    expect(getByTestId("article")).toHaveTextContent(data.cards[0].date.toString());

    expect(getByTestId("article")).not.toHaveTextContent(data.cards[1].name);
    expect(getByTestId("article")).not.toHaveTextContent(data.cards[1].description);
    expect(getByTestId("article")).not.toHaveTextContent(data.cards[1].text);
    expect(getByTestId("article")).not.toHaveTextContent(data.cards[1].time.toString());
    expect(getByTestId("article")).not.toHaveTextContent(data.cards[1].date.toString());
  });
});
