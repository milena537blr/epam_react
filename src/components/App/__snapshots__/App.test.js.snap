// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> snapshot test 1`] = `
<App
  data={
    Object {
      "cards": Array [
        Object {
          "date": 2007,
          "description": "American drama",
          "genre": "Dramas",
          "id": 1,
          "name": "Vampire diaries",
          "text": "The Vampire Diaries is an American supernatural teen drama television series developed by Kevin Williamson and Julie Plec, based on the popular book series of the same name written by L. J. Smith.",
          "time": 145,
        },
        Object {
          "date": 2014,
          "description": "Drama/Comedy-drama",
          "genre": "Drama",
          "id": 2,
          "name": "The Fault in Our Stars",
          "text": "Hazel Grace Lancaster (Shailene Woodley), a 16-year-old cancer patient, meets and falls in love with Gus Waters (Ansel Elgort), a similarly afflicted teen from her cancer support group. Hazel feels that Gus really understands her. ",
          "time": 133,
        },
      ],
    }
  }
>
  <ErrorBoundary>
    <header
      className="header"
    >
      <div
        className="overlay"
      />
      <section
        className="container"
      >
        <Box
          align="space-between"
          direction="horizontal"
          inline={false}
          marginBottom={8}
          verticalAlign="middle"
        >
          <div
            className="root horizontal spaceBetween middle"
            style={
              Object {
                "height": "undefined",
                "margin": "undefined",
                "marginBottom": 48,
                "marginLeft": "undefined",
                "marginRight": "undefined",
                "marginTop": "undefined",
                "maxHeight": "undefined",
                "maxWidth": "undefined",
                "minHeight": "undefined",
                "minWidth": "undefined",
                "padding": "undefined",
                "paddingBottom": "undefined",
                "paddingLeft": "undefined",
                "paddingRight": "undefined",
                "paddingTop": "undefined",
                "width": "undefined",
              }
            }
          >
            <Logo>
              <a
                className="logo"
                href="/"
              >
                netflixroulette
              </a>
            </Logo>
            <Button
              area-label="Display search area"
              color="white"
              data-testid="search-switcher"
              onHandleSearchClick={[Function]}
              size="large"
              text="SEARCH"
            >
              <button
                className="button large white"
                onClick={[Function]}
                type="submit"
              >
                SEARCH
              </button>
            </Button>
          </div>
        </Box>
        <Search
          onSearchTextChange={[Function]}
          searchText=""
        >
          <div
            className="search"
            data-testid="search"
          >
            <form>
              <fieldset>
                <legend
                  className="searchTitle header__searchTitle"
                >
                  Find your movie
                </legend>
                <Box
                  direction="horizontal"
                  inline={false}
                  marginBottom={4}
                >
                  <div
                    className="root horizontal"
                    style={
                      Object {
                        "height": "undefined",
                        "margin": "undefined",
                        "marginBottom": 24,
                        "marginLeft": "undefined",
                        "marginRight": "undefined",
                        "marginTop": "undefined",
                        "maxHeight": "undefined",
                        "maxWidth": "undefined",
                        "minHeight": "undefined",
                        "minWidth": "undefined",
                        "padding": "undefined",
                        "paddingBottom": "undefined",
                        "paddingLeft": "undefined",
                        "paddingRight": "undefined",
                        "paddingTop": "undefined",
                        "width": "undefined",
                      }
                    }
                  >
                    <input
                      aria-label="Search"
                      aria-labelledby="searchButton"
                      className="searchInput"
                      name="search"
                      onChange={[Function]}
                      placeholder="Quentin Tarantino"
                      type="text"
                      value=""
                    />
                  </div>
                </Box>
                <Box
                  align="space-between"
                  direction="horizontal"
                  inline={false}
                  verticalAlign="middle"
                >
                  <div
                    className="root horizontal spaceBetween middle"
                    style={
                      Object {
                        "height": "undefined",
                        "margin": "undefined",
                        "marginBottom": "undefined",
                        "marginLeft": "undefined",
                        "marginRight": "undefined",
                        "marginTop": "undefined",
                        "maxHeight": "undefined",
                        "maxWidth": "undefined",
                        "minHeight": "undefined",
                        "minWidth": "undefined",
                        "padding": "undefined",
                        "paddingBottom": "undefined",
                        "paddingLeft": "undefined",
                        "paddingRight": "undefined",
                        "paddingTop": "undefined",
                        "width": "undefined",
                      }
                    }
                  >
                    <Box
                      align="space-between"
                      direction="horizontal"
                      inline={false}
                      verticalAlign="middle"
                    >
                      <div
                        className="root horizontal spaceBetween middle"
                        style={
                          Object {
                            "height": "undefined",
                            "margin": "undefined",
                            "marginBottom": "undefined",
                            "marginLeft": "undefined",
                            "marginRight": "undefined",
                            "marginTop": "undefined",
                            "maxHeight": "undefined",
                            "maxWidth": "undefined",
                            "minHeight": "undefined",
                            "minWidth": "undefined",
                            "padding": "undefined",
                            "paddingBottom": "undefined",
                            "paddingLeft": "undefined",
                            "paddingRight": "undefined",
                            "paddingTop": "undefined",
                            "width": "undefined",
                          }
                        }
                      >
                        <div
                          className="searchLabel"
                        >
                          search by
                        </div>
                        <Box
                          direction="horizontal"
                          inline={false}
                          marginRight={2}
                        >
                          <div
                            className="root horizontal"
                            style={
                              Object {
                                "height": "undefined",
                                "margin": "undefined",
                                "marginBottom": "undefined",
                                "marginLeft": "undefined",
                                "marginRight": 12,
                                "marginTop": "undefined",
                                "maxHeight": "undefined",
                                "maxWidth": "undefined",
                                "minHeight": "undefined",
                                "minWidth": "undefined",
                                "padding": "undefined",
                                "paddingBottom": "undefined",
                                "paddingLeft": "undefined",
                                "paddingRight": "undefined",
                                "paddingTop": "undefined",
                                "width": "undefined",
                              }
                            }
                          >
                            <Button
                              color="red"
                              size="medium"
                              text="title"
                            >
                              <button
                                className="button medium red"
                                type="submit"
                              >
                                title
                              </button>
                            </Button>
                          </div>
                        </Box>
                        <Box
                          direction="horizontal"
                          inline={false}
                          marginRight={2}
                        >
                          <div
                            className="root horizontal"
                            style={
                              Object {
                                "height": "undefined",
                                "margin": "undefined",
                                "marginBottom": "undefined",
                                "marginLeft": "undefined",
                                "marginRight": 12,
                                "marginTop": "undefined",
                                "maxHeight": "undefined",
                                "maxWidth": "undefined",
                                "minHeight": "undefined",
                                "minWidth": "undefined",
                                "padding": "undefined",
                                "paddingBottom": "undefined",
                                "paddingLeft": "undefined",
                                "paddingRight": "undefined",
                                "paddingTop": "undefined",
                                "width": "undefined",
                              }
                            }
                          >
                            <Button
                              color="gray"
                              size="medium"
                              text="genre"
                            >
                              <button
                                className="button medium gray"
                                type="submit"
                              >
                                genre
                              </button>
                            </Button>
                          </div>
                        </Box>
                      </div>
                    </Box>
                    <Button
                      color="red"
                      id="searchButton"
                      size="large"
                      text="SEARCH"
                    >
                      <button
                        className="button large red"
                        type="submit"
                      >
                        SEARCH
                      </button>
                    </Button>
                  </div>
                </Box>
              </fieldset>
            </form>
          </div>
        </Search>
      </section>
    </header>
    <section
      className="panel"
    >
      <Box
        align="space-between"
        className="container"
        direction="horizontal"
        inline={false}
        verticalAlign="middle"
      >
        <div
          className="root container horizontal spaceBetween middle"
          style={
            Object {
              "height": "undefined",
              "margin": "undefined",
              "marginBottom": "undefined",
              "marginLeft": "undefined",
              "marginRight": "undefined",
              "marginTop": "undefined",
              "maxHeight": "undefined",
              "maxWidth": "undefined",
              "minHeight": "undefined",
              "minWidth": "undefined",
              "padding": "undefined",
              "paddingBottom": "undefined",
              "paddingLeft": "undefined",
              "paddingRight": "undefined",
              "paddingTop": "undefined",
              "width": "undefined",
            }
          }
        >
          <div>
            7 movies found
          </div>
          <div>
            <span>
              Sort by 
            </span>
            <span>
              release date 
            </span>
            <span>
              rating
            </span>
          </div>
        </div>
      </Box>
    </section>
    <main
      className="main"
    >
      <div
        className="container"
      >
        <Box
          align="center"
          className="wrapper"
          direction="horizontal"
          inline={false}
        >
          <div
            className="root wrapper horizontal center"
            style={
              Object {
                "height": "undefined",
                "margin": "undefined",
                "marginBottom": "undefined",
                "marginLeft": "undefined",
                "marginRight": "undefined",
                "marginTop": "undefined",
                "maxHeight": "undefined",
                "maxWidth": "undefined",
                "minHeight": "undefined",
                "minWidth": "undefined",
                "padding": "undefined",
                "paddingBottom": "undefined",
                "paddingLeft": "undefined",
                "paddingRight": "undefined",
                "paddingTop": "undefined",
                "width": "undefined",
              }
            }
          >
            <Card
              card={
                Object {
                  "date": 2007,
                  "genre": "Dramas",
                  "id": 1,
                  "name": "Vampire diaries",
                }
              }
              cardClick={[Function]}
              key="1"
            >
              <figure
                className="card"
              >
                <button
                  area-label="Vampire diaries"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                >
                  <img
                    alt="Vampire diaries"
                    className="image"
                    src="test-file-stub"
                  />
                </button>
                <figcaption>
                  <Box
                    align="space-between"
                    direction="horizontal"
                    inline={false}
                    verticalAlign="middle"
                  >
                    <div
                      className="root horizontal spaceBetween middle"
                      style={
                        Object {
                          "height": "undefined",
                          "margin": "undefined",
                          "marginBottom": "undefined",
                          "marginLeft": "undefined",
                          "marginRight": "undefined",
                          "marginTop": "undefined",
                          "maxHeight": "undefined",
                          "maxWidth": "undefined",
                          "minHeight": "undefined",
                          "minWidth": "undefined",
                          "padding": "undefined",
                          "paddingBottom": "undefined",
                          "paddingLeft": "undefined",
                          "paddingRight": "undefined",
                          "paddingTop": "undefined",
                          "width": "undefined",
                        }
                      }
                    >
                      <div
                        className="name"
                      >
                        Vampire diaries
                      </div>
                      <div
                        className="date"
                      >
                        2007
                      </div>
                    </div>
                  </Box>
                  <div
                    className="genre"
                  >
                    Dramas
                  </div>
                </figcaption>
              </figure>
            </Card>
            <Card
              card={
                Object {
                  "date": 2014,
                  "genre": "Drama",
                  "id": 2,
                  "name": "The Fault in Our Stars",
                }
              }
              cardClick={[Function]}
              key="2"
            >
              <figure
                className="card"
              >
                <button
                  area-label="The Fault in Our Stars"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                >
                  <img
                    alt="The Fault in Our Stars"
                    className="image"
                    src="test-file-stub"
                  />
                </button>
                <figcaption>
                  <Box
                    align="space-between"
                    direction="horizontal"
                    inline={false}
                    verticalAlign="middle"
                  >
                    <div
                      className="root horizontal spaceBetween middle"
                      style={
                        Object {
                          "height": "undefined",
                          "margin": "undefined",
                          "marginBottom": "undefined",
                          "marginLeft": "undefined",
                          "marginRight": "undefined",
                          "marginTop": "undefined",
                          "maxHeight": "undefined",
                          "maxWidth": "undefined",
                          "minHeight": "undefined",
                          "minWidth": "undefined",
                          "padding": "undefined",
                          "paddingBottom": "undefined",
                          "paddingLeft": "undefined",
                          "paddingRight": "undefined",
                          "paddingTop": "undefined",
                          "width": "undefined",
                        }
                      }
                    >
                      <div
                        className="name"
                      >
                        The Fault in Our Stars
                      </div>
                      <div
                        className="date"
                      >
                        2014
                      </div>
                    </div>
                  </Box>
                  <div
                    className="genre"
                  >
                    Drama
                  </div>
                </figcaption>
              </figure>
            </Card>
          </div>
        </Box>
      </div>
    </main>
    <footer
      className="footer"
    >
      <div
        className="container"
      >
        <Logo>
          <a
            className="logo"
            href="/"
          >
            netflixroulette
          </a>
        </Logo>
      </div>
    </footer>
  </ErrorBoundary>
</App>
`;
